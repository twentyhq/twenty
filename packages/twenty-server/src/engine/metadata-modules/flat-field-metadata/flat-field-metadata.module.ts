import { Module } from '@nestjs/common';

import { FeatureFlagModule } from 'src/engine/core-modules/feature-flag/feature-flag.module';

import { FlatFieldMetadataTypeValidatorService } from './services/flat-field-metadata-type-validator.service';
import { FlatFieldMetadataValidatorService } from './services/flat-field-metadata-validator.service';

@Module({
  imports: [FeatureFlagModule],
  providers: [
    FlatFieldMetadataValidatorService,
    FlatFieldMetadataTypeValidatorService,
  ],
  exports: [
    FlatFieldMetadataValidatorService,
    FlatFieldMetadataTypeValidatorService,
  ],
})
export class FlatFieldMetadataModule {}
