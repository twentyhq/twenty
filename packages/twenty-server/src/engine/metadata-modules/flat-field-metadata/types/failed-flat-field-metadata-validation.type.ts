import { isDefined } from 'twenty-shared/utils';

import {
  FieldMetadataException,
  FieldMetadataExceptionCode,
} from 'src/engine/metadata-modules/field-metadata/field-metadata.exception';
import { type ObjectMetadataException } from 'src/engine/metadata-modules/object-metadata/object-metadata.exception';
import { type InvalidMetadataException } from 'src/engine/metadata-modules/utils/exceptions/invalid-metadata.exception';

export type FailedFlatFieldMetadataValidationExceptions =
  | FieldMetadataException
  | ObjectMetadataException
  | InvalidMetadataException;

export type FlatFieldMetadataValidator<T> = {
  validator: (value: T) => boolean;
  message: string;
};

export const runFlatFieldMetadataValidator = <T>(
  elementToValidate: T,
  { message, validator }: FlatFieldMetadataValidator<T>,
): FailedFlatFieldMetadataValidationExceptions | undefined => {
  const validationFailed = validator(elementToValidate);

  if (validationFailed) {
    return new FieldMetadataException(
      message,
      FieldMetadataExceptionCode.INVALID_FIELD_INPUT,
      {
        userFriendlyMessage: message,
      },
    );
  }
};

export const runFlatFieldMetadataValidators = <T>(
  elementToValidate: T,
  validators: FlatFieldMetadataValidator<T>[],
): FailedFlatFieldMetadataValidationExceptions[] =>
  validators
    .map((validator) =>
      runFlatFieldMetadataValidator(elementToValidate, validator),
    )
    .filter(isDefined);
