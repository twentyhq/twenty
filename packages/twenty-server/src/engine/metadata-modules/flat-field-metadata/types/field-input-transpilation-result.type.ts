import { type FailedFlatFieldMetadataValidationExceptions } from 'src/engine/metadata-modules/flat-field-metadata/types/failed-flat-field-metadata-validation.type';

export type FailedFieldInputTranspilation = {
  status: 'fail';
  error: FailedFlatFieldMetadataValidationExceptions;
};
export type SuccessfulFieldInputTranspilation<T> = {
  status: 'success';
  result: T;
};
export type FieldInputTranspilationResult<T> =
  | FailedFieldInputTranspilation
  | SuccessfulFieldInputTranspilation<T>;
