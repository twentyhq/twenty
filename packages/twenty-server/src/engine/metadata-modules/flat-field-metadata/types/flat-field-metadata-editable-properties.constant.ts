import { FLAT_FIELD_METADATA_EDITABLE_PROPERTIES } from "src/engine/metadata-modules/flat-field-metadata/constants/flat-field-metadata-editable-properties.constant";

export type FlatFieldMetadataEditableProperties =
  (typeof FLAT_FIELD_METADATA_EDITABLE_PROPERTIES)[number];
