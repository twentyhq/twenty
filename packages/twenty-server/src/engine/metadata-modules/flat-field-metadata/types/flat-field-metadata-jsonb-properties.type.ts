import { type FLAT_FIELD_METADATA_JSONB_PROPERTIES } from 'src/engine/metadata-modules/flat-field-metadata/constants/flat-field-metadata-jsonb-properties.constant';

export type FlatFieldMetadataEntityJsonbProperties =
  (typeof FLAT_FIELD_METADATA_JSONB_PROPERTIES)[number];
