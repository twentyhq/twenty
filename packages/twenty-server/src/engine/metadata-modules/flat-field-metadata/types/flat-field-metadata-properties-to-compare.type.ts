import { type FLAT_FIELD_METADATA_PROPERTIES_TO_COMPARE } from 'src/engine/metadata-modules/flat-field-metadata/constants/flat-field-metadata-properties-to-compare.constant';

export type FlatFieldMetadataPropertiesToCompare =
  (typeof FLAT_FIELD_METADATA_PROPERTIES_TO_COMPARE)[number];
