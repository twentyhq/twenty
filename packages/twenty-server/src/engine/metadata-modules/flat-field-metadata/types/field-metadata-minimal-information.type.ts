import { type FlatFieldMetadata } from 'src/engine/metadata-modules/flat-field-metadata/types/flat-field-metadata.type';

export type FieldMetadataMinimalInformation = Pick<
  FlatFieldMetadata,
  'id' | 'objectMetadataId' | 'name'
>;
