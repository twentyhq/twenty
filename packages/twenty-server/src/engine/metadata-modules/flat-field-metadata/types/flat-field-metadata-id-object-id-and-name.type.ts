import { FlatFieldMetadata } from "src/engine/metadata-modules/flat-field-metadata/types/flat-field-metadata.type";

export type FlatFieldMetadataIdObjectIdAndName = Pick<
  FlatFieldMetadata,
  'id' | 'objectMetadataId' | 'name'
>;