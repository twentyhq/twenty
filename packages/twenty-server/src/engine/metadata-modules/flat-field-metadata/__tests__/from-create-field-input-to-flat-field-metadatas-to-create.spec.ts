import { FieldMetadataType } from 'twenty-shared/types';

import { type CreateFieldInput } from 'src/engine/metadata-modules/field-metadata/dtos/create-field.input';
import { RelationType } from 'src/engine/metadata-modules/field-metadata/interfaces/relation-type.interface';
import { FieldInputTranspilationResult } from 'src/engine/metadata-modules/flat-field-metadata/types/field-input-transpilation-result.type';
import { FlatFieldMetadata } from 'src/engine/metadata-modules/flat-field-metadata/types/flat-field-metadata.type';
import { fromCreateFieldInputToFlatFieldMetadatasToCreate } from 'src/engine/metadata-modules/flat-field-metadata/utils/from-create-field-input-to-flat-field-metadatas-to-create.util';
import { FLAT_OBJECT_METADATA_MAPS_MOCKS } from 'src/engine/metadata-modules/flat-object-metadata-maps/mocks/flat-object-metadata-maps.mock';
import { type FlatObjectMetadataMaps } from 'src/engine/metadata-modules/flat-object-metadata-maps/types/flat-object-metadata-maps.type';
import { replaceFlatObjectMetadataInFlatObjectMetadataMapsOrThrow } from 'src/engine/metadata-modules/flat-object-metadata-maps/utils/replace-flat-object-metadata-in-flat-object-metadata-maps-or-throw.util';
import { COMPANY_FLAT_OBJECT_MOCK } from 'src/engine/metadata-modules/flat-object-metadata/__mocks__/company-flat-object.mock';
import { PET_FLAT_OBJECT_MOCK } from 'src/engine/metadata-modules/flat-object-metadata/__mocks__/pet-flat-object.mock';
import { ROCKET_FLAT_OBJECT_MOCK } from 'src/engine/metadata-modules/flat-object-metadata/__mocks__/rocket-flat-object.mock';
import { extractRecordIdsAndDatesAsExpectAny } from 'test/utils/extract-record-ids-and-dates-as-expect-any';
import { EachTestingContext } from 'twenty-shared/testing';

describe('fromCreateFieldInputToFlatFieldMetadatasToCreate', () => {
  const mockWorkspaceId = 'mock-workspace-id';
  const existingFlatObjectMetadataMaps = [
    COMPANY_FLAT_OBJECT_MOCK,
    ROCKET_FLAT_OBJECT_MOCK,
    PET_FLAT_OBJECT_MOCK,
  ].reduce((flatObjectMetadataMaps, flatObjectMetadata) => {
    return replaceFlatObjectMetadataInFlatObjectMetadataMapsOrThrow({
      flatObjectMetadata: {
        ...flatObjectMetadata,
        flatFieldMetadatas: [],
      },
      flatObjectMetadataMaps,
    });
  }, FLAT_OBJECT_METADATA_MAPS_MOCKS);

  const testCases: EachTestingContext<{
    input: {
      rawCreateFieldInput: CreateFieldInput &
        Required<Pick<CreateFieldInput, 'morphRelationsCreationPayload'>>;
      workspaceId: string;
      existingFlatObjectMetadataMaps: FlatObjectMetadataMaps;
    };
    expected: Partial<FieldInputTranspilationResult<FlatFieldMetadata[]>>;
  }>[] = [
    {
      only: true,
      title:
        'should create morph relation field metadata with valid input on rocket object to pet object',
      context: {
        input: {
          rawCreateFieldInput: {
            name: 'newField',
            label: 'newFieldLabel',
            description: 'new field description',
            type: FieldMetadataType.MORPH_RELATION,
            objectMetadataId: ROCKET_FLAT_OBJECT_MOCK.id,
            workspaceId: mockWorkspaceId,
            morphRelationsCreationPayload: [
              {
                type: RelationType.ONE_TO_MANY,
                targetObjectMetadataId: PET_FLAT_OBJECT_MOCK.id,
                targetFieldLabel: 'Pet',
                targetFieldIcon: 'IconBuilding',
              },
              {
                type: RelationType.ONE_TO_MANY,
                targetObjectMetadataId: COMPANY_FLAT_OBJECT_MOCK.id,
                targetFieldLabel: 'Company',
                targetFieldIcon: 'IconBuilding',
              },
            ],
          },
          workspaceId: mockWorkspaceId,
          existingFlatObjectMetadataMaps,
        },
        expected: {
          status: 'success',
        },
      },
    },
  ];

  test.each(testCases)('$title', async ({ context: { input, expected } }) => {
    const result =
      await fromCreateFieldInputToFlatFieldMetadatasToCreate(input);
    console.log(result);
    expect(result.status).toEqual(expected.status);
    if (expected.status === 'success' && result.status === 'success') {
      expect(result.result.length).toBe(
        input.rawCreateFieldInput.morphRelationsCreationPayload.length * 2,
      );
      expect(result).toMatchSnapshot(
        extractRecordIdsAndDatesAsExpectAny(result),
      );
    } else {
      expect(result).toMatchSnapshot(
        extractRecordIdsAndDatesAsExpectAny(result),
      );
    }
  });
});
