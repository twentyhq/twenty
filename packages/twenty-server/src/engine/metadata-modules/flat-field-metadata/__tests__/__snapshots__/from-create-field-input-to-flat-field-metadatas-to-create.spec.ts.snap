// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromCreateFieldInputToFlatFieldMetadatasToCreate should create morph relation field metadata with valid input on rocket object to pet object 1`] = `
{
  "result": [
    {
      "createdAt": Any<ClockDate>,
      "defaultValue": null,
      "description": "new field description",
      "flatRelationTargetFieldMetadata": {
        "createdAt": Any<ClockDate>,
        "defaultValue": null,
        "description": null,
        "flatRelationTargetFieldMetadata": {
          "createdAt": Any<ClockDate>,
          "defaultValue": null,
          "description": "new field description",
          "flatRelationTargetFieldMetadata": null,
          "flatRelationTargetObjectMetadata": {
            "createdAt": Any<String>,
            "description": null,
            "duplicateCriteria": null,
            "flatFieldMetadatas": [],
            "flatIndexMetadatas": [],
            "icon": "IconCat",
            "id": Any<String>,
            "imageIdentifierFieldMetadataId": null,
            "isActive": true,
            "isAuditLogged": true,
            "isCustom": true,
            "isLabelSyncedWithName": false,
            "isRemote": false,
            "isSearchable": true,
            "isSystem": false,
            "isUIReadOnly": false,
            "labelIdentifierFieldMetadataId": Any<String>,
            "labelPlural": "Pets",
            "labelSingular": "Pet",
            "namePlural": "pets",
            "nameSingular": "pet",
            "shortcut": null,
            "standardId": null,
            "standardOverrides": null,
            "targetTableName": "DEPRECATED",
            "uniqueIdentifier": "d34e0f07-1b8c-4de0-938e-599cf05e1f7f",
            "updatedAt": Any<String>,
            "workspaceId": Any<String>,
          },
          "icon": "IconRelationOneToMany",
          "id": Any<String>,
          "isActive": true,
          "isCustom": true,
          "isLabelSyncedWithName": false,
          "isNullable": true,
          "isSystem": false,
          "isUIReadOnly": false,
          "isUnique": null,
          "label": "newFieldLabel",
          "name": "newField",
          "objectMetadataId": Any<String>,
          "options": null,
          "relationTargetFieldMetadataId": Any<String>,
          "relationTargetObjectMetadataId": Any<String>,
          "settings": {
            "relationType": "ONE_TO_MANY",
          },
          "standardId": null,
          "standardOverrides": null,
          "type": "MORPH_RELATION",
          "uniqueIdentifier": "f3bad14d-9faf-408a-aea5-9803c836e532",
          "updatedAt": Any<ClockDate>,
          "workspaceId": Any<String>,
        },
        "flatRelationTargetObjectMetadata": {
          "createdAt": Any<String>,
          "description": "A rocket",
          "duplicateCriteria": null,
          "flatFieldMetadatas": [],
          "flatIndexMetadatas": [],
          "icon": "IconRocket",
          "id": Any<String>,
          "imageIdentifierFieldMetadataId": null,
          "isActive": true,
          "isAuditLogged": true,
          "isCustom": true,
          "isLabelSyncedWithName": false,
          "isRemote": false,
          "isSearchable": true,
          "isSystem": false,
          "isUIReadOnly": false,
          "labelIdentifierFieldMetadataId": Any<String>,
          "labelPlural": "Rockets",
          "labelSingular": "Rocket",
          "namePlural": "rockets",
          "nameSingular": "rocket",
          "shortcut": null,
          "standardId": null,
          "standardOverrides": null,
          "targetTableName": "DEPRECATED",
          "uniqueIdentifier": "d78ec657-74a4-4652-a350-1f44ff62970a",
          "updatedAt": Any<String>,
          "workspaceId": Any<String>,
        },
        "icon": "IconBuilding",
        "id": Any<String>,
        "isActive": true,
        "isCustom": true,
        "isLabelSyncedWithName": false,
        "isNullable": true,
        "isSystem": false,
        "isUIReadOnly": false,
        "isUnique": null,
        "label": "Pet",
        "name": "pet",
        "objectMetadataId": Any<String>,
        "options": null,
        "relationTargetFieldMetadataId": Any<String>,
        "relationTargetObjectMetadataId": Any<String>,
        "settings": {
          "joinColumnName": "petId",
          "onDelete": "SET_NULL",
          "relationType": "MANY_TO_ONE",
        },
        "standardId": null,
        "standardOverrides": null,
        "type": "RELATION",
        "uniqueIdentifier": "c3c5cb9f-d970-487a-8fce-59439f06d81c",
        "updatedAt": Any<ClockDate>,
        "workspaceId": Any<String>,
      },
      "flatRelationTargetObjectMetadata": {
        "createdAt": Any<String>,
        "description": null,
        "duplicateCriteria": null,
        "flatFieldMetadatas": [],
        "flatIndexMetadatas": [],
        "icon": "IconCat",
        "id": Any<String>,
        "imageIdentifierFieldMetadataId": null,
        "isActive": true,
        "isAuditLogged": true,
        "isCustom": true,
        "isLabelSyncedWithName": false,
        "isRemote": false,
        "isSearchable": true,
        "isSystem": false,
        "isUIReadOnly": false,
        "labelIdentifierFieldMetadataId": Any<String>,
        "labelPlural": "Pets",
        "labelSingular": "Pet",
        "namePlural": "pets",
        "nameSingular": "pet",
        "shortcut": null,
        "standardId": null,
        "standardOverrides": null,
        "targetTableName": "DEPRECATED",
        "uniqueIdentifier": "d34e0f07-1b8c-4de0-938e-599cf05e1f7f",
        "updatedAt": Any<String>,
        "workspaceId": Any<String>,
      },
      "icon": "IconRelationOneToMany",
      "id": Any<String>,
      "isActive": true,
      "isCustom": true,
      "isLabelSyncedWithName": false,
      "isNullable": true,
      "isSystem": false,
      "isUIReadOnly": false,
      "isUnique": null,
      "label": "newFieldLabel",
      "name": "newField",
      "objectMetadataId": Any<String>,
      "options": null,
      "relationTargetFieldMetadataId": Any<String>,
      "relationTargetObjectMetadataId": Any<String>,
      "settings": {
        "relationType": "ONE_TO_MANY",
      },
      "standardId": null,
      "standardOverrides": null,
      "type": "MORPH_RELATION",
      "uniqueIdentifier": "f3bad14d-9faf-408a-aea5-9803c836e532",
      "updatedAt": Any<ClockDate>,
      "workspaceId": Any<String>,
    },
    {
      "createdAt": Any<ClockDate>,
      "defaultValue": null,
      "description": null,
      "flatRelationTargetFieldMetadata": {
        "createdAt": Any<ClockDate>,
        "defaultValue": null,
        "description": "new field description",
        "flatRelationTargetFieldMetadata": null,
        "flatRelationTargetObjectMetadata": {
          "createdAt": Any<String>,
          "description": null,
          "duplicateCriteria": null,
          "flatFieldMetadatas": [],
          "flatIndexMetadatas": [],
          "icon": "IconCat",
          "id": Any<String>,
          "imageIdentifierFieldMetadataId": null,
          "isActive": true,
          "isAuditLogged": true,
          "isCustom": true,
          "isLabelSyncedWithName": false,
          "isRemote": false,
          "isSearchable": true,
          "isSystem": false,
          "isUIReadOnly": false,
          "labelIdentifierFieldMetadataId": Any<String>,
          "labelPlural": "Pets",
          "labelSingular": "Pet",
          "namePlural": "pets",
          "nameSingular": "pet",
          "shortcut": null,
          "standardId": null,
          "standardOverrides": null,
          "targetTableName": "DEPRECATED",
          "uniqueIdentifier": "d34e0f07-1b8c-4de0-938e-599cf05e1f7f",
          "updatedAt": Any<String>,
          "workspaceId": Any<String>,
        },
        "icon": "IconRelationOneToMany",
        "id": Any<String>,
        "isActive": true,
        "isCustom": true,
        "isLabelSyncedWithName": false,
        "isNullable": true,
        "isSystem": false,
        "isUIReadOnly": false,
        "isUnique": null,
        "label": "newFieldLabel",
        "name": "newField",
        "objectMetadataId": Any<String>,
        "options": null,
        "relationTargetFieldMetadataId": Any<String>,
        "relationTargetObjectMetadataId": Any<String>,
        "settings": {
          "relationType": "ONE_TO_MANY",
        },
        "standardId": null,
        "standardOverrides": null,
        "type": "MORPH_RELATION",
        "uniqueIdentifier": "f3bad14d-9faf-408a-aea5-9803c836e532",
        "updatedAt": Any<ClockDate>,
        "workspaceId": Any<String>,
      },
      "flatRelationTargetObjectMetadata": {
        "createdAt": Any<String>,
        "description": "A rocket",
        "duplicateCriteria": null,
        "flatFieldMetadatas": [],
        "flatIndexMetadatas": [],
        "icon": "IconRocket",
        "id": Any<String>,
        "imageIdentifierFieldMetadataId": null,
        "isActive": true,
        "isAuditLogged": true,
        "isCustom": true,
        "isLabelSyncedWithName": false,
        "isRemote": false,
        "isSearchable": true,
        "isSystem": false,
        "isUIReadOnly": false,
        "labelIdentifierFieldMetadataId": Any<String>,
        "labelPlural": "Rockets",
        "labelSingular": "Rocket",
        "namePlural": "rockets",
        "nameSingular": "rocket",
        "shortcut": null,
        "standardId": null,
        "standardOverrides": null,
        "targetTableName": "DEPRECATED",
        "uniqueIdentifier": "d78ec657-74a4-4652-a350-1f44ff62970a",
        "updatedAt": Any<String>,
        "workspaceId": Any<String>,
      },
      "icon": "IconBuilding",
      "id": Any<String>,
      "isActive": true,
      "isCustom": true,
      "isLabelSyncedWithName": false,
      "isNullable": true,
      "isSystem": false,
      "isUIReadOnly": false,
      "isUnique": null,
      "label": "Pet",
      "name": "pet",
      "objectMetadataId": Any<String>,
      "options": null,
      "relationTargetFieldMetadataId": Any<String>,
      "relationTargetObjectMetadataId": Any<String>,
      "settings": {
        "joinColumnName": "petId",
        "onDelete": "SET_NULL",
        "relationType": "MANY_TO_ONE",
      },
      "standardId": null,
      "standardOverrides": null,
      "type": "RELATION",
      "uniqueIdentifier": "c3c5cb9f-d970-487a-8fce-59439f06d81c",
      "updatedAt": Any<ClockDate>,
      "workspaceId": Any<String>,
    },
    {
      "createdAt": Any<ClockDate>,
      "defaultValue": null,
      "description": "new field description",
      "flatRelationTargetFieldMetadata": {
        "createdAt": Any<ClockDate>,
        "defaultValue": null,
        "description": null,
        "flatRelationTargetFieldMetadata": {
          "createdAt": Any<ClockDate>,
          "defaultValue": null,
          "description": "new field description",
          "flatRelationTargetFieldMetadata": null,
          "flatRelationTargetObjectMetadata": {
            "createdAt": Any<String>,
            "description": "A company",
            "duplicateCriteria": null,
            "flatFieldMetadatas": [],
            "flatIndexMetadatas": [],
            "icon": "IconBuildingSkyscraper",
            "id": Any<String>,
            "imageIdentifierFieldMetadataId": null,
            "isActive": true,
            "isAuditLogged": true,
            "isCustom": false,
            "isLabelSyncedWithName": false,
            "isRemote": false,
            "isSearchable": true,
            "isSystem": false,
            "isUIReadOnly": false,
            "labelIdentifierFieldMetadataId": Any<String>,
            "labelPlural": "Companies",
            "labelSingular": "Company",
            "namePlural": "companies",
            "nameSingular": "company",
            "shortcut": "C",
            "standardId": Any<String>,
            "standardOverrides": null,
            "targetTableName": "DEPRECATED",
            "uniqueIdentifier": "20202020-b374-4779-a561-80086cb2e17f",
            "updatedAt": Any<String>,
            "workspaceId": Any<String>,
          },
          "icon": "IconRelationOneToMany",
          "id": Any<String>,
          "isActive": true,
          "isCustom": true,
          "isLabelSyncedWithName": false,
          "isNullable": true,
          "isSystem": false,
          "isUIReadOnly": false,
          "isUnique": null,
          "label": "newFieldLabel",
          "name": "newField",
          "objectMetadataId": Any<String>,
          "options": null,
          "relationTargetFieldMetadataId": Any<String>,
          "relationTargetObjectMetadataId": Any<String>,
          "settings": {
            "relationType": "ONE_TO_MANY",
          },
          "standardId": null,
          "standardOverrides": null,
          "type": "MORPH_RELATION",
          "uniqueIdentifier": "b9bff059-75ca-44ec-b59a-4e98cd8a83b2",
          "updatedAt": Any<ClockDate>,
          "workspaceId": Any<String>,
        },
        "flatRelationTargetObjectMetadata": {
          "createdAt": Any<String>,
          "description": "A rocket",
          "duplicateCriteria": null,
          "flatFieldMetadatas": [],
          "flatIndexMetadatas": [],
          "icon": "IconRocket",
          "id": Any<String>,
          "imageIdentifierFieldMetadataId": null,
          "isActive": true,
          "isAuditLogged": true,
          "isCustom": true,
          "isLabelSyncedWithName": false,
          "isRemote": false,
          "isSearchable": true,
          "isSystem": false,
          "isUIReadOnly": false,
          "labelIdentifierFieldMetadataId": Any<String>,
          "labelPlural": "Rockets",
          "labelSingular": "Rocket",
          "namePlural": "rockets",
          "nameSingular": "rocket",
          "shortcut": null,
          "standardId": null,
          "standardOverrides": null,
          "targetTableName": "DEPRECATED",
          "uniqueIdentifier": "d78ec657-74a4-4652-a350-1f44ff62970a",
          "updatedAt": Any<String>,
          "workspaceId": Any<String>,
        },
        "icon": "IconBuilding",
        "id": Any<String>,
        "isActive": true,
        "isCustom": true,
        "isLabelSyncedWithName": false,
        "isNullable": true,
        "isSystem": false,
        "isUIReadOnly": false,
        "isUnique": null,
        "label": "Company",
        "name": "company",
        "objectMetadataId": Any<String>,
        "options": null,
        "relationTargetFieldMetadataId": Any<String>,
        "relationTargetObjectMetadataId": Any<String>,
        "settings": {
          "joinColumnName": "companyId",
          "onDelete": "SET_NULL",
          "relationType": "MANY_TO_ONE",
        },
        "standardId": null,
        "standardOverrides": null,
        "type": "RELATION",
        "uniqueIdentifier": "db6d59dc-02da-47b8-add4-92f4ca7320b2",
        "updatedAt": Any<ClockDate>,
        "workspaceId": Any<String>,
      },
      "flatRelationTargetObjectMetadata": {
        "createdAt": Any<String>,
        "description": "A company",
        "duplicateCriteria": null,
        "flatFieldMetadatas": [],
        "flatIndexMetadatas": [],
        "icon": "IconBuildingSkyscraper",
        "id": Any<String>,
        "imageIdentifierFieldMetadataId": null,
        "isActive": true,
        "isAuditLogged": true,
        "isCustom": false,
        "isLabelSyncedWithName": false,
        "isRemote": false,
        "isSearchable": true,
        "isSystem": false,
        "isUIReadOnly": false,
        "labelIdentifierFieldMetadataId": Any<String>,
        "labelPlural": "Companies",
        "labelSingular": "Company",
        "namePlural": "companies",
        "nameSingular": "company",
        "shortcut": "C",
        "standardId": Any<String>,
        "standardOverrides": null,
        "targetTableName": "DEPRECATED",
        "uniqueIdentifier": "20202020-b374-4779-a561-80086cb2e17f",
        "updatedAt": Any<String>,
        "workspaceId": Any<String>,
      },
      "icon": "IconRelationOneToMany",
      "id": Any<String>,
      "isActive": true,
      "isCustom": true,
      "isLabelSyncedWithName": false,
      "isNullable": true,
      "isSystem": false,
      "isUIReadOnly": false,
      "isUnique": null,
      "label": "newFieldLabel",
      "name": "newField",
      "objectMetadataId": Any<String>,
      "options": null,
      "relationTargetFieldMetadataId": Any<String>,
      "relationTargetObjectMetadataId": Any<String>,
      "settings": {
        "relationType": "ONE_TO_MANY",
      },
      "standardId": null,
      "standardOverrides": null,
      "type": "MORPH_RELATION",
      "uniqueIdentifier": "b9bff059-75ca-44ec-b59a-4e98cd8a83b2",
      "updatedAt": Any<ClockDate>,
      "workspaceId": Any<String>,
    },
    {
      "createdAt": Any<ClockDate>,
      "defaultValue": null,
      "description": null,
      "flatRelationTargetFieldMetadata": {
        "createdAt": Any<ClockDate>,
        "defaultValue": null,
        "description": "new field description",
        "flatRelationTargetFieldMetadata": null,
        "flatRelationTargetObjectMetadata": {
          "createdAt": Any<String>,
          "description": "A company",
          "duplicateCriteria": null,
          "flatFieldMetadatas": [],
          "flatIndexMetadatas": [],
          "icon": "IconBuildingSkyscraper",
          "id": Any<String>,
          "imageIdentifierFieldMetadataId": null,
          "isActive": true,
          "isAuditLogged": true,
          "isCustom": false,
          "isLabelSyncedWithName": false,
          "isRemote": false,
          "isSearchable": true,
          "isSystem": false,
          "isUIReadOnly": false,
          "labelIdentifierFieldMetadataId": Any<String>,
          "labelPlural": "Companies",
          "labelSingular": "Company",
          "namePlural": "companies",
          "nameSingular": "company",
          "shortcut": "C",
          "standardId": Any<String>,
          "standardOverrides": null,
          "targetTableName": "DEPRECATED",
          "uniqueIdentifier": "20202020-b374-4779-a561-80086cb2e17f",
          "updatedAt": Any<String>,
          "workspaceId": Any<String>,
        },
        "icon": "IconRelationOneToMany",
        "id": Any<String>,
        "isActive": true,
        "isCustom": true,
        "isLabelSyncedWithName": false,
        "isNullable": true,
        "isSystem": false,
        "isUIReadOnly": false,
        "isUnique": null,
        "label": "newFieldLabel",
        "name": "newField",
        "objectMetadataId": Any<String>,
        "options": null,
        "relationTargetFieldMetadataId": Any<String>,
        "relationTargetObjectMetadataId": Any<String>,
        "settings": {
          "relationType": "ONE_TO_MANY",
        },
        "standardId": null,
        "standardOverrides": null,
        "type": "MORPH_RELATION",
        "uniqueIdentifier": "b9bff059-75ca-44ec-b59a-4e98cd8a83b2",
        "updatedAt": Any<ClockDate>,
        "workspaceId": Any<String>,
      },
      "flatRelationTargetObjectMetadata": {
        "createdAt": Any<String>,
        "description": "A rocket",
        "duplicateCriteria": null,
        "flatFieldMetadatas": [],
        "flatIndexMetadatas": [],
        "icon": "IconRocket",
        "id": Any<String>,
        "imageIdentifierFieldMetadataId": null,
        "isActive": true,
        "isAuditLogged": true,
        "isCustom": true,
        "isLabelSyncedWithName": false,
        "isRemote": false,
        "isSearchable": true,
        "isSystem": false,
        "isUIReadOnly": false,
        "labelIdentifierFieldMetadataId": Any<String>,
        "labelPlural": "Rockets",
        "labelSingular": "Rocket",
        "namePlural": "rockets",
        "nameSingular": "rocket",
        "shortcut": null,
        "standardId": null,
        "standardOverrides": null,
        "targetTableName": "DEPRECATED",
        "uniqueIdentifier": "d78ec657-74a4-4652-a350-1f44ff62970a",
        "updatedAt": Any<String>,
        "workspaceId": Any<String>,
      },
      "icon": "IconBuilding",
      "id": Any<String>,
      "isActive": true,
      "isCustom": true,
      "isLabelSyncedWithName": false,
      "isNullable": true,
      "isSystem": false,
      "isUIReadOnly": false,
      "isUnique": null,
      "label": "Company",
      "name": "company",
      "objectMetadataId": Any<String>,
      "options": null,
      "relationTargetFieldMetadataId": Any<String>,
      "relationTargetObjectMetadataId": Any<String>,
      "settings": {
        "joinColumnName": "companyId",
        "onDelete": "SET_NULL",
        "relationType": "MANY_TO_ONE",
      },
      "standardId": null,
      "standardOverrides": null,
      "type": "RELATION",
      "uniqueIdentifier": "db6d59dc-02da-47b8-add4-92f4ca7320b2",
      "updatedAt": Any<ClockDate>,
      "workspaceId": Any<String>,
    },
  ],
  "status": "success",
}
`;
